File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RewriteRuleCompositionResult.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/** Auditable result of a composed rewrite sequence. */
public record RewriteRuleCompositionResult(SemanticPlan plan,List<PlanRewriteRuleResult> applications,double totalCostImpact,boolean terminatedNormally,List<RewriteRuleCandidate> selectionHistory,List<ProviderAwareRewriteRuleCandidate> providerSelectionHistory) {
    public RewriteRuleCompositionResult {applications=List.copyOf(applications);selectionHistory=selectionHistory==null?List.of():List.copyOf(selectionHistory);providerSelectionHistory=providerSelectionHistory==null?List.of():List.copyOf(providerSelectionHistory);}
    public RewriteRuleCompositionResult(SemanticPlan plan,List<PlanRewriteRuleResult> applications,double totalCostImpact,boolean terminatedNormally){this(plan,applications,totalCostImpact,terminatedNormally,List.of(),List.of());}
    public List<String> appliedRules(){return applications.stream().map(PlanRewriteRuleResult::ruleName).toList();}
    public boolean changed(){return !applications.isEmpty();}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/RewriteRuleCompositionResult.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Auditable result of a composed rewrite sequence.</summary>
public sealed record RewriteRuleCompositionResult(
    SemanticPlan Plan,
    IReadOnlyList<PlanRewriteRuleResult> Applications,
    double TotalCostImpact,
    bool TerminatedNormally,
    IReadOnlyList<RewriteRuleCandidate>? SelectionHistory = null,
    IReadOnlyList<ProviderAwareRewriteRuleCandidate>? ProviderSelectionHistory = null)
{
    public IReadOnlyList<string> AppliedRules => Applications.Select(x => x.RuleName).ToArray();
    public bool Changed => Applications.Count != 0;
    public IReadOnlyList<RewriteRuleCandidate> Candidates => SelectionHistory ?? [];
    public IReadOnlyList<ProviderAwareRewriteRuleCandidate> ProviderCandidates => ProviderSelectionHistory ?? [];
}